File: max_heap/max_heap.rb
class MaxHeap

	def initialize
		@heap = []
		@size = 0
	end

	def insert value
		@heap << value
		@size++
			heapify
	end

	# Ensures the node at positition follows the heap property.
	# Assumes the children are valid heaps
	def heapify position
		largest = determine_largest position

		if largest != position
			swap = @heap[position]
			@heap[position] = @heap[largest]
			@heap[largest] = swap
		end

		return self

	end

	def pop
		if @size < 1
			return nil
		end

		max = @heap[0]
		@heap[0] = @heap[@size - 1]

		heapify 1

		return max
	end

	def peek
		return @heap[0]
	end

	def sort
	end

	def left_child_position position
		position * 2
	end

	def right_child_position position
		position * 2 + 1
	end

	def parent position
		(position / 2).floor
	end

	def determine_largest position
		left_position = left_child_position
		right_position = right_child_position

		largest = if left_position <= @size && @heap[left_position - 1] > @heap[position - 1]
					  left_position
				  else
					  position
				  end

		largest = if right_position <= @size && @heap[right_posiion - 1] > largest
					  right_position
				  else
					  largest
				  end
		largest
	end

end

File: max_heap/c/src/max_heap.c
#include "max_heap.h"

static Heap * heapify(Heap *heap);

Heap *
heapCreate() {
	Heap *heap = malloc(sizeof(Heap));

	heap -> array = arrayCreate();
	heap -> size  = 0;

	return heap;
}

Heap *
heapInsert(Heap *heap, void *value) {
	array *array = heap -> array;
	arrayPushTail(array, value);
	heap -> size = heap -> array -> size;
	heapify(heap);
	return heap;
}

Heap *
heapify(Heap *heap) {
	if (heap -> size == 1) return heap;

	array *array = heap -> array;
	int size = heap -> size;
	int parent = getParent(size);

	int leftPosition = getLeftChild(size / 2);
	int largest = size;
	if (leftPosition <= size) {
		int left = *(int*)getLeftChild(size)
	}
	int rightPosition = getRightChild(size);

	void *left = arrayRetrieve(array, leftPosition - 1);
	void *right = arrayRetrieve(array, rightPosition - 1);

	

	return heap;
}

int
getLeftChild(int parent) {
	return parent * 2;
}

int
getRightChild(int parent) {
	return parent * 2 + 1;
}

int
getParent(int child) {
	if (child <=1 ) return child;

	return child / 2;
}
